src/ops/base.rs

```diff
@@ -1,39 +1,221 @@
+use serde_json;
+use snowflake::ProcessUniqueId;
+use buf_redux::BufWriter;
+use buf_redux::strategy::WhenFull;
 use std::collections::HashMap;
+use std::fs;
+use std::fs::{File, OpenOptions};
+use std::path::PathBuf;
+use std::time::{Duration, Instant};
+use time;
+
+// 4k buffered log.
+const LOG_BUFFER_CAPACITY: usize = 4 * 1024;
+
+// We store at most this many write records before flushing to disk.
+const BUFFERED_WRITES_CAPACITY: usize = 512;
+
+// We spend at least this many milliseconds without flushing write records to disk.
+const BUFFERED_WRITES_FLUSH_INTERVAL_MS: u64 = 1000;
 
 /// Base is used to represent the root nodes of the distributary data flow graph.
 ///
 /// These nodes perform no computation, and their job is merely to persist all received updates and
 /// forward them to interested downstream operators. A base node should only be sent updates of the
 /// type corresponding to the node's type.
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 pub struct Base {
+    buffered_writes: Option<Records>,
+    durability: BaseDurabilityLevel,
+    durable_log: Option<BufWriter<File, WhenFull>>,
+    durable_log_path: Option<PathBuf>,
+    global_address: Option<NodeAddress>,
+    last_flushed_at: Option<Instant>,
```

Comment:
Yes, naming `Base` nodes at creation time would be ideal, and it would also make it easier for a recovery manager to restore `Base` nodes -- since now an external entity can keep track of the global_id -> Base node mapping, it knows where to fetch the corresponding durable log for a given node.

How does global id generation currently work right now?  Could we expose that as a thread-safe singleton somewhere, so that every time we call `Base::new` it gets assigned an unique pseudo random number (LCG, xorshift, etc)?